File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
  public int[] topKFrequent(int[] nums, int k) {
    Map<Integer, Integer> counts = new HashMap<>();

    for (int num : nums) {
      counts.put(num, counts.getOrDefault(num, 0) + 1);
    }

    // Buckets = [ 0=[], 1=[], 2=[], ...] & here 0, 1, 2 are frequencies 
    List<Integer>[] buckets = new List[nums.length + 1];

    for (Map.Entry<Integer, Integer> entry : counts.entrySet()) {
      int num = entry.getKey();
      int freq = entry.getValue();

      if (buckets[freq] == null)
        buckets[freq] = new ArrayList<>();

      buckets[freq].add(num);
    }

    int[] ans = new int[k];
    int idx = 0;

    for (int freq = buckets.length - 1; freq > 0 && idx < k; freq--) {
      if (buckets[freq] == null)
        continue;

      for (int num : buckets[freq]) {
        ans[idx++] = num;

        if (idx == k)
          return ans;
      }
    }
    return ans;
  }
}

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
      # from collections import Counter
      # freq = Counter(nums)
      # buckets = [[] for _ in range(len(nums) + 1)]

      # for x, f in freq.items():
      #     buckets[f].append(x)

      # ans = []
      # for f in range(len(buckets) - 1, 0, -1):
      #     ans.extend(buckets[f])
      #     if len(ans) >= k:
      #         return ans[:k]
      return [x[0] for x in Counter(nums).most_common(k)]